proposals/readonly-parameters.md

```diff
@@ -0,0 +1,115 @@
+
+C# users have [long requested](https://github.com/dotnet/csharplang/issues/188) the ability to mark both locals and parameters as `readonly`. The design team has somewhat resisted this for two reasons:
+
+* The view that `readonly` on parameters and locals would be an attractive nuisance more than a helpful addition.
+* Indecision on what a succinct syntax for locals would be to minimize the nuisance part of the first objection.
+
+However, the addition of primary constructor parameters changes that calculus, at least for parameters. A significant piece of feedback from the initial preview is that users would like to be able to
+ensure that primary constructor parameters are not modified. The scope of such parameters is much larger, so the danger of accidental modification is much higher. We therefore propose allowing `readonly`
+as a parameter modifier.
+
+This proposal makes a few assumptions about `readonly` locals as part of it:
+
+* A future design would allow `readonly` as a local modifier.
+* That future design might allow a shorthand for `readonly var`, or may say that `readonly var` is not allowed, and the separate shorthand is required for a `readonly` type-inferred local. What that shorthand
+  is (`val`, `let`, `const`, or some other keyword) is beyond the scope of this proposal.
+
+These assumptions allow us to presume that to fully spell out readonlyness for a parameter or local requires a modifier, and that modifier is `readonly`. In places where types can be inferred, we offer a
+shorthand that combines the meanings, but otherwise `readonly` is required.
+
+## Detailed design
+[design]: #detailed-design
+
+### Syntax
+
+We modify [section 15.6.2.1](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/classes.md#15621-general) of the C# specification with the following new definition for `parameter_modifier`:
+
+```antlr
+parameter_modifier
+    : parameter_mode_modifier
+    | 'this'
+    | 'readonly'
+    ;
+```
+
+The spec is not yet updated to include the draft C# 12 specification for [`ref readonly` parameters](https://github.com/dotnet/csharplang/blob/main/proposals/ref-readonly-parameters.md), but we will restrict
+`readonly` such that if a parameter is both `readonly` and `ref readonly`, the `readonly` must appear on the left side of the `ref readonly`, so that `readonly ref readonly` is permitted, but `ref readonly readonly`
+is not.
+
+### Semantics
+
+For a `readonly` parameter, the compiler will issue an error when it is assigned to or taken as a mutable `lvalue`. This means that a `readonly` parameter cannot be passed by `out` or `ref`, but can be passed
+by value, `ref readonly`, or `in`.
+
+#### `partial` methods
+
+For `partial` methods, we allow the implementing `partial` method declaration to add the `readonly` modifier to a parameter if the defining `partial` method did not. If the defining partial `method` included the
+`readonly` modifier, the implementing `partial` method must also include it.
```

Comment:
as mentioned offline, my personal preference would be to *not* allow this.  But it's not a blocker for me.